umbra-js/src/classes/Umbra.ts

```diff
@@ -410,23 +413,37 @@ export class Umbra {
     acceptor: string,
     token: string,
     sponsor: string,
-    sponsorFee: BigNumberish
+    sponsorFee: BigNumberish,
+    hook: string = AddressZero,
+    data: string | null = null
   ) {
     // Validate addresses
     acceptor = getAddress(acceptor);
     sponsor = getAddress(sponsor);
     token = getAddress(token);
+    hook = getAddress(hook);
 
     // Validate chainId
     if (typeof chainId !== 'number' || !Number.isInteger(chainId)) {
       throw new Error(`Invalid chainId provided in chainConfig. Got '${chainId}'`);
     }
 
+    let dataArg;
+    if (data === null) {
+      // Empty data should passed as an empty array to encode
+      dataArg = [];
+    } else if (typeof data !== 'string' || !data.startsWith('0x')) {
+      // Validate the data string
+      throw new Error('Data string must be in hex form with 0x prefix');
+    } else {
+      dataArg = data;
+    }
```

Comment:
A few things on this section:

1. Ah, so maybe the `[]` over `""` is not just a web3 thing! Maybe just set the default data value to `[]` so we can skip the first check and just initialize with `let dataArg = data`? Guessing you didn't do that because TS complained about `[]` as a type, though

2. Prefer to use `!isHexString` here to be more robust instead of checking the first two characters are 0x

3. Seems error message should be `Data input must be null or a string in hex form with 0x prefix`

3. Seems we could tidy things up a bit to remove some branches and only have one `if` condition that checks for a valid hex string. If you can't use `[]` as the default value, and are feeling emboldened to use some ternary, you could even just do the default input as `data: string = "0x"` then in the method do:

```typescript
if (typeof data !== 'string' || !isHexString(data)) {
  throw new Error('...')
}
const dataArg = data === '0x' ? [] : data;
```

Reply:
> Ah, so maybe the [] over "" is not just a web3 thing! Maybe just set the default data value to [] so we can skip the first check and just initialize with let dataArg = data? Guessing you didn't do that because TS complained about [] as a type, though

See above— same error for both web3 and ethers.

> Prefer to use !isHexString here to be more robust instead of checking the first two characters are 0x
>
> Seems error message should be Data input must be null or a string in hex form with 0x prefix


Good call, on both. Will update.

> If you can't use [] as the default value

So, I'm not actually sure if I can use `[]` because I didn't try. As mentioned, I really disliked the underlying libraries requiring `[]` for null data, so I wrote our public method intentionally to accept `null` or a hex string.

> are feeling emboldened to use some ternary, you could even just do the default input as data: string = "0x" then in the method do

I find this harder to read and less clear as a public interface 😬 . I think the caller should be able to pass `null` if they want, and if they don't, it should default to no data.

Reply:
> So, I'm not actually sure if I can use [] because I didn't try. As mentioned, I really disliked the underlying libraries requiring [] for null data, so I wrote our public method intentionally to accept null or a hex string.

Give our new understanding of why they both use `[]`, this might be more acceptable now and we may want to allow it

Reply:
Updated with ecec84a6d4a55c2a17e07cf2403260c38947b232